src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -415,9 +449,55 @@ private async Task<bool> TryReadPrefaceAsync()
             {
                 if (!readableBuffer.IsEmpty)
                 {
-                    if (ParsePreface(readableBuffer, out consumed, out examined))
+                    if (state.HasFlag(ReadPrefaceState.Preface))
+                    {
+                        if (readableBuffer.Length >= ClientPreface.Length)
+                        {
+                            if (IsPreface(readableBuffer, out consumed, out examined))
+                            {
+                                return true;
+                            }
+                            else
+                            {
+                                state ^= ReadPrefaceState.Preface;
+                            }
+                        }
+                    }
+
+                    if (state.HasFlag(ReadPrefaceState.Http1x))
+                    {
+                        if (ParseHttp1x(readableBuffer, out var detectedVersion))
```

Comment:
Nit: It's not super obvious from the name that `ParseHttp1x` returns `true` when the `MaxRequestLineSize` is exceeded before an `\r` could be found. Maybe `ParseHttp1x` should just take the current `ReadPrefaceState` and return the updated one or something.